### PR TITLE
httpbakery: add PublicKeyRing

### PR DESCRIPTION
**Pull request by rogpeppe**
Also move the existing PublicKeyForLocation function and tests into the same files.

### CODE REVIEW

File: httpbakery/keyring.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	}
+	k, err := kr.cache.PublicKeyForLocation(loc)
+	if err == nil {
+		return k, nil
+	}
+	k, err = PublicKeyForLocation(kr.client, loc)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	kr.cache.AddPublicKeyForLocation(loc, true, k)
```
Comment (alesstimec) [inline httpbakery/keyring.go]: err = kr.cache.AddPublicKeyForLocation(...)
if err != nil {
}

File: httpbakery/keyring_test.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	jujutesting.LoggingSuite
+}
+
+var _ = gc.Suite(&KeyringSuite{})
+
+func (s *KeyringSuite) TestCachePrepopulated(c *gc.C) {
+	cache := bakery.NewPublicKeyRing()
+	key, err := bakery.GenerateKey()
+	c.Assert(err, gc.IsNil)
+	cache.AddPublicKeyForLocation("https://0.1.2.3/", true, &key.Public)
```
Comment (alesstimec) [inline httpbakery/keyring_test.go:29]: check the returned error

**Comment by alesstimec** [comment]
LGTM after the two comments above are addressed..

File: httpbakery/keyring.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+func NewPublicKeyRing(client *http.Client, cache *bakery.PublicKeyRing) *PublicKeyRing {
+	if cache == nil {
+		cache = bakery.NewPublicKeyRing()
+	}
+	if client == nil {
+		client = http.DefaultClient
+	}
+	return &PublicKeyRing{
+		client: client,
+		cache:  cache,
```
Comment (alesstimec) [inline httpbakery/keyring.go:29]: i know bool defaults to false, but it might be nice to be explicit here
allowInsecure: false,

Comment (rogpeppe) [inline httpbakery/keyring.go:29]: I think leaving it as omitted is reasonable. The sense of the boolean is that way for a reason.

File: httpbakery/keyring.go (hunk 1 above)
Comment (rogpeppe) [inline httpbakery/keyring.go]: done (even though it actually can never happen)

File: httpbakery/keyring_test.go (hunk 2 above)
Comment (rogpeppe) [inline httpbakery/keyring_test.go:29]: if it fails, the test will fail, so I don't care much.

File: httpbakery/keyring.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		// loc is an invalid URL which we have already checked.
+		return nil, errgo.Notef(err, "cannot cache discharger URL %q", loc)
+	}
+	return k, nil
+}
+
+// PublicKeyForLocation returns the public key from a macaroon
+// discharge server running at the given location URL.
+// Note that this is insecure if an http: URL scheme is used.
+func PublicKeyForLocation(client *http.Client, url string) (*bakery.PublicKey, error) {
```
Comment (mhilton) [inline httpbakery/keyring.go:80]: It might be nice if you could use nil for the client parameter and have it use http.DefaultClient

**Comment by mhilton** [comment]
:+1: LGTM with a small thought.

File: httpbakery/keyring.go (hunk 4 above)
Comment (rogpeppe) [inline httpbakery/keyring.go:80]: done

**Pull request merged**